### ISSUE TITLE
`nnls` example code runs out of memory

### ISSUE THREAD

**Issue by MLanghof** [librosa/librosa#1634]
**Describe the bug**

`nnls` attempts to make a (transient) > 2 GB memory allocation when running the `nnls` example code, which fails on my machine.

**To Reproduce**

Run the example code from https://librosa.org/doc/main/generated/librosa.util.nnls.html#librosa.util.nnls:

![image](https://user-images.githubusercontent.com/6524684/209738507-cb5d7c3e-e61f-4220-8805-bc875a8cc369.png)

**Discussion**

The code appears to try and circumvent memory problems by processing the input in blocks depending on `MAX_MEM_BLOCK`, but it doesn't appear to correctly judge the quadratic growth of the work array `wa` deep in the fortran wrapper code.

A 2 GB memory budget is probably not unreasonable to expect nowadays (I am on a more, let's say, "portable" setup atm), but in my original use case it actually tried to allocate 16 GB. I don't think that's reasonable, and it's unfortunate that the code path for avoiding it exists but doesn't actually prevent it.

If `MAX_MEM_BLOCK` were tunable, I could probably mess with it a bit to make it work. (I tried various monkey-patching in this direction, but I'm probably not doing it right or it's not possible. Can I adjust `MAX_MEM_BLOCK` somehow from the outside? If so, how?)

**Expected behavior**

No out-of-memory error, or a user-facing tuning parameter for working around this issue.

**Software versions**

```
Windows-8.1-6.3.9600-SP0
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:16:33) [MSC v.1929 64 bit (AMD64)]
NumPy 1.22.4
SciPy 1.8.1
librosa 0.9.2
INSTALLED VERSIONS
------------------
python: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:16:33) [MSC v.1929 64 bit (AMD64)]

librosa: 0.9.2

audioread: 3.0.0
numpy: 1.22.4
scipy: 1.8.1
sklearn: 1.1.2
joblib: 1.2.0
decorator: 5.1.1
soundfile: 0.11.0
resampy: 0.4.2
numba: 0.56.4
pooch: v1.6.0
packaging: 22.0

numpydoc: None
sphinx: None
sphinx_rtd_theme: None
sphinx_multiversion: None
sphinx_gallery: None
mir_eval: None
ipython: None
sphinxcontrib-svg2pdfconverter: None
pytest: None
pytest-mpl: None
pytest-cov: None
matplotlib: 3.6.2
samplerate: None
soxr: None
contextlib2: None
presets: None
```

**Additional context**

I actually ran into this via `librosa.feature.inverse.mel_to_audio`, which runs into the above when given more than a handful of frames at a time. And unlike when calling `nnls` directly, I don't see a good way to piece together the results of separate "few frames at a time" calls meaningfully there.

**Comment by bmcfee** [comment]
> but in my original use case it actually tried to allocate 16 GB.

That's pretty strange - I wonder if it's something upstream from us (eg within scipy or blas), or possibly the python gc not running?

For the record, when I profile this in memray:

```python
%%memray_flamegraph --trace-python-allocators --leaks

S_recover = librosa.util.nnls(mel_basis, M)
```
I get stats:
```
Start time: 2022-12-27 20:51:54.323000
End time: 2022-12-27 20:51:54.803000
Total number of allocations: 1328991
Total number of frames seen: 418
Peak memory usage: 2.3 GB
Python allocator: pymalloc
```
Quite far from the 16GB you're describing.  Digging into the flamegraph, I see the allocation for `wa` that you describe and it indeed takes the majority (2.1GB).

You can indeed override `librosa.util.MAX_MEMORY_BLOCK` to whatever size you want.  This is an undocumented feature, but it will do what you expect.  I played around with this on my dev machine and for both much smaller and much larger sizes, I get the same behavior in memray.  Your mileage may vary, but I suspect this is more of a garbage collector issue than anything else.

<details> <summary> Dev environment dump </summary>

```python
INSTALLED VERSIONS
------------------
python: 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:45:29) 
[GCC 10.4.0]

librosa: 0.10.0.dev0

audioread: 3.0.0
numpy: 1.23.5
scipy: 1.9.3
sklearn: 1.2.0
joblib: 1.2.0
decorator: 5.1.1
soundfile: 0.11.0
resampy: 0.4.2
numba: 0.56.4
pooch: v1.6.0
packaging: 22.0

numpydoc: 1.5.0
sphinx: 5.3.0
sphinx_rtd_theme: 1.1.1
sphinx_multiversion: 0.2.4
sphinx_gallery: 0.11.1
mir_eval: 0.6
ipython: None
sphinxcontrib-svg2pdfconverter: None
pytest: 7.2.0
pytest-mpl: None
pytest-cov: None
matplotlib: 3.6.2
samplerate: 0.1.0
soxr: 0.3.3
presets: 0.1.3
```


</details>

**Comment by MLanghof** [comment]
Hi, thanks for the quick reply!

The higher allocation sizes do not happen with the exact example code indeed, but the point is that they happen easily (_despite_ the code trying to contain memory usage). Here is a very slight variation of the example code for demonstration purposes (double the fft size, and 8x as many frames in the spectrogram):

```
    y, sr = librosa.load(librosa.ex('trumpet'), duration=3)
    S = np.abs(librosa.stft(y, n_fft=4096))
    M = librosa.feature.melspectrogram(S=S, sr=sr, power=1
    M = np.tile(M, 8)
    mel_basis = librosa.filters.mel(sr=sr, n_fft=4096, n_mels=M.shape[0])
    S_recover = librosa.util.nnls(mel_basis, M)
```

This results in:
```
Unable to allocate 32.4 GiB for an array with shape (4350606867,) and data type float64`
```
(It actually goes into the blockwise code here for me, but it just happens to choose `n_columns` such that there is a 32 GB allocation in there.)

In short:
- The example code makes a 2 GB allocation. This might be "not a bug" territory, but at least in my case this causes it to fail.
- `nnls` tries to limit memory usage (for good reason) but this is not effective in the general case. I believe this is a bug.

It's not a garbage collector issue, my current system is simply too old/weak to easily satisfy multi-GB allocations. I do not observe a leak or anything - when the allocation succeeds, the computation is over and the memory released before I see it in any kind of resource monitoring.

**Comment by bmcfee** [comment]
> It's not a garbage collector issue, my current system is simply too old/weak to easily satisfy multi-GB allocations.

I see - I was misinterpreting this as a redundant allocation not being cleaned up across the block loop iterations, and had not realized that you had modified the example code.



> * The example code makes a 2 GB allocation. This might be "not a bug" territory, but at least in my case this causes it to fail.

I think this might be unavoidable, as the allocations are happening deeper into the call stack (ie inside LFBGS).  Not sure what we can do here.



> * `nnls` tries to limit memory usage (for good reason) but this is not effective in the general case. I believe this is a bug.

I think this is a bit of a gray area.  The blocking here was never intended as a contract—thou shalt not use more than X memory—just as a way to reduce memory usage without sacrificing too much in terms of running time.  This also occurs in STFT for similar reasons.



> Edit: Finally figured out how to set `MAX_MEMORY_BLOCK` in a way that affects `nnls` - it's just `librosa.util._nnls.MAX_MEM_BLOCK = 2**8` (or whichever size works) and can be changed at any time.

Good catch - I had forgotten that the NNLS module will get its own copy of the blocking parameter.  But your solution ought to work and is unlikely to break any time soon.  I'll also note that if you wanted finer grained control of the solver, you can control fmin_lbfgs by pass-through arguments (eg to reduce the rank of the approximate inverse hessian).

All that said, it's probably worth discussing opening up some API to control memory usage.  I've been hesitant to do this before because A) this is python and we really can't control these things too carefully, and B) it's never been a real problem before.

**Comment by MLanghof** [comment]
Thank you for the comments!

For my particular case that clears things up I guess. If it's niche enough of a problem then extending the API may not be worthwhile, at least if the above workaround is available for the time being (but that's of course for you to weigh). Maybe this discussion even helps a lost soul in the future :)

**Comment by MLanghof** [comment]
Another note:

> you can control fmin_lbfgs by pass-through arguments

You actually can't when using `mel_to_audio`, because `kwargs` will also be passed to `mel`, which will complain about unused parameters ¯\\\_(ツ)_/¯

**Comment by bmcfee** [comment]
Good point - it wouldn't be too hard to unroll a step and code mel_to_stft directly (and then griffin-lim the result), but if you're hacking nnls already it's a moot point.
